### PR TITLE
Configure network addresses from sam4l

### PR DESCRIPTION
**Pull request by hudson-ayers**
### Pull Request Overview

This PR modifies main.rs for the Imix such that each Imix generates a 16 bit short address for its source MAC address from the serial number of the sam4l on the board. It also adds a utility function to generate link local IPv6 addresses from an EUI 64 (mac address), and uses this to set one of the possible ipv6 source addresses for the device (by adding it to the interface list). This PR goes hand-in-hand with an upcoming PR that will add udp send/rx apps for use at the Sensys tutorial.

### Testing Strategy

I tested this by flashing multiple Imixes with the app and confirming that each generated a unique, but consistent address, and that I could use these addresses to send/receive UDP packets as expected.

### TODO or Help Wanted

Remove commits that are redundant from PRs now merged into master

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

File: chips/sam4l/src/serial_num.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    /// as a u64 rather than a byte array
+    pub fn get_lower_64(&self) -> u64 {
+        let full_num = self.regs.serial_num;
+        let mut lower_64 = 0;
+        // Below could use transmute, but didn't want to add unsafe code
+        for i in 7..15 {
+            lower_64 = lower_64 + ((full_num[i] as u64) << ((14-i)*8));
+        }
+        lower_64
+    }
```
Comment (niklasad1) [inline chips/sam4l/src/serial_num.rs]: Consider making this code a little more `rusty` something like should be equivalent:

```rust
   full_num.iter()
            .rev()
            .take(8)
            .enumerate()
            .fold(0_64, |sum, (i, &val)| {
                (val as u64) << i*8
            })
```

File: boards/imix/src/main.rs [hunk 2]
```diff
@@ -411,11 +375,9 @@ pub unsafe fn reset_handler() {
...
     // For now, assign the MAC address on the device as simply a 16-bit short address which represents
     // the last 16 bits of the serial number of the sam4l for this device.
-    // In the future, we could generate the MAC address by hasing the full 120-bit serial number
-    let serial_num: SerialNum = SerialNum::new();
-    //let serial_flash_upper = serial_num.regs.upper64.get();
-    let serial_flash_lower = serial_num.regs.lower64.get();
-    let serial_num_bottom_16 = (serial_flash_lower & 0x000000000000ffff) as u16;
+    // In the future, we could generate the MAC address by hashing the full 120-bit serial number
+    let serial_num: sam4l::serial_num::SerialNum = sam4l::serial_num::SerialNum::new();
+    let serial_num_bottom_16 = (serial_num.get_lower_64() & 0x000000000000ffff) as u16;
```
Comment (niklasad1) [inline boards/imix/src/main.rs]: Use underscores on hex numbers for readability i.e, ```0x0000_0000_0000_ffff```

**Comment by hudson-ayers** [comment]
I just opened a new PR bc I was struggling with rebasing because I merged multiple commits out of order from how they were merged onto master, and because Amit amended his commit after I cherry-picked it and before he put it into master, so I found it was easier to just open a new branch off master and cherry-pick commits from this branch accordingly.

**Comment by hudson-ayers** [comment]
@niklasad1 Thanks for your comments! I will address them on the new PR

**Pull request closed without merging**